src/main/kotlin/com/google/androidstudiopoet/input/NewInputExample.json

```diff
@@ -0,0 +1,103 @@
+{
+  //Project configuration is wrapped in another object to help us deal with versioning
+  "inputVersion": "0.0.1",
+  "projectConfig": {
+    "projectName": "genny",
+    "root": "./modules/",
+    "buildSystemConfig": {
+      //Optional param, but it is possible to specify what exactly build system setup should be used
+    },
+    "moduleConfigs": [
+      //Even though there are only 3 blocks here, in total 11 modules would be generated.
+      //One other thing, I believe that we should forbid dependencies withing one block, only interblock dependencies are allowed.
+      {
+        //Andorid Application Module
+        "moduleType": "ANDROID",
+        "moduleNamePrefix": "app",
+        "isApplication": true,
+
+        "numActivities": 8,
+        "resourcesConfig": {
+          "stringCount": 10,
+          "imageCount": 9,
+          "layoutCount": 7
+        },
+
+        "javaPackageCount": 3,
+        "javaClassCount": 8,
+        "javaMethodCount": 2000,
+
+        "kotlinPackageCount": 3,
+        "kotlinClassCount": 5,
+        "kotlinMethodCount": 1000,
+
+        "dependencies": [
```

Comment:
What do you think on also having explicit dependencies outside ModuleConfigs? Following this structure:

class ExplicitDependency {
  val from : List<String>
  val to : List<String>
  val type : String?
}

The format accepted by "from" and "to" would be a list of strings, each representing a prefix (meaning all the modules of that ModuleConfig) or a specific module.

An example is like this:

"dependencies": [{"from": ["prefixA", "prefixB5"], "to": ["prefixC", "prefixB1"]}, {"from": ["prefixA"], "to": ["prefixB0"], "type": "api"}]

That would create the following:

- ((all modules in prefixA) and prefixB5)) depend on ((all modules in prefixC) and prefixB1) (with default type (implementation)
- (all modules in prefixA) depend on prefixB0 for api

We can let the type for later since that has other implications (for example, how the code dependencies should be generated). This format (a list of prefixes and modules) can be used in other places to describe a set of modules and can be reused in topologies.

While thinking on how input processing will be implemented, at the end, all of this will be used to generate a list of explicit module names that will be added to the build.gradle files. I think both formats can easily be adapted to use the same structure that will hold those strings. Right now it is a List<Dependency> that contains pairs of integers. This structure needs to change to support dependencies between different ModuleConfig. What do you think of changing this list to Map<String, List<Dependency>>? The key will be the module name while the list holds the modules it depends on. We can also change the structure of Dependency since now the "from" will be redundant and also add the dependency type ("api", "implementation", etc.).

One advantage of using a map instead of a list is that we will not require to filter on every module since the map will give us that in constant time, also improving performance (in the worst case, this list can have N*(N-1)/2 dependencies, causing the process of adding dependencies into build files to be cubic on the number of modules, while using a map is quadratic).

Reply:
> Right now it is a List<Dependency> that contains pairs of integers.

That is true, but it will change completely with the new format, each `ModuleConfig` will have a List<DependencyConfig> where it refers other ModuleConfig by prefix. https://github.com/NikitaKozlov/android-studio-poet/blob/71be1842b8c112d3e979ae083926df09a3dfef1d/src/main/kotlin/com/google/androidstudiopoet/input/DependencyConfig.kt#L3

It is possible to express with new format everything you listed ("api" and "implementation" is easy to add), although it will be scattered through different `ModuleConfig`s:
``` json

{
  "inputVersion": "0.0.1",
  "projectConfig": {
    "projectName": "genny",
    "root": "./modules/",
    "moduleConfigs": [
      {
        "moduleCount": 5,
        "moduleNamePrefix": "prefixA",
        "dependencies": [
          {
            "type": "BULK",
            "moduleNamePrefix": "prefixC",
            "moduleCount": 5
          },
          {
            "type": "BULK",
            "moduleNamePrefix": "prefixB1",
            "moduleCount": 5
          },
          {
            "type": "BULK",
            "moduleNamePrefix": "prefixB0",
            "moduleCount": 5,
            "dependencyType": "api"
          }
        ]
      },
      {
        "moduleCount": 5,
        "moduleNamePrefix": "prefixB5",
        "dependencies": [
          {
            "type": "BULK",
            "moduleNamePrefix": "prefixC",
            "moduleCount": 5
          },
          {
            "type": "BULK",
            "moduleNamePrefix": "prefixB1",
            "moduleCount": 5
          }
        ]
      },
      {
        "moduleCount": 5,
        "moduleNamePrefix": "prefixC"
      },
      {
        "moduleCount": 5,
        "moduleNamePrefix": "prefixB1"
      },
      {
        "moduleCount": 5,
        "moduleNamePrefix": "prefixB0"
      }
    ]
  }
}

```
Because each `ModuleConfig` explicitly shows what other `ModuleConfig`s it depends on then there'll be no more filtering. 

I think we can first implement one way of doing things and after we play with it we can reevaluate it. If it would be too verbose, or it wouldn't fit some of the needs (for example topologies) we can add `ExplicitDependency` later.

I'm not sure how do you want to use `ExplicitDependency` for topologies, can you please explain it? 

Unfortunately module name is not enough for generating dependency, we also need method that needs to be called and resources that need to be refered, currently figuring out those things could be computationally expensive.

Reply:
Can you please explain as well the problem you are trying to solve? Then it would be easier for me to agree with you our offer some alternatives.